### DISCUSSION TITLE
FTP-based firmware update

### DISCUSSION THREAD

**Discussion by matth-x** [OpenEVSE/openevse_esp32_firmware#786] [Ideas]
Hi!

I was wondering if there are good use cases for an FTP integration within OpenEVSE. Although HTTP seems to be the very preferred way of file transfers on microcontrollers, there may be some online services which only support FTP. Were there any related request so far?

The background is that I'm planning on updating my OpenEVSE fork / PR with my new [FTP client](https://github.com/matth-x/MicroFtp) because it's the only really supported file transfer method of OCPP backends. Either the integration will be local to the OCPP module or I would extend the http_update module accordingly. Any opinions on this?

**Comment by jeremypoulter** [comment]
I haven't seen any other requests for FTP but don't see any reason not to add it. I would probably create a new module to do the FTP update. The code the http_update file is just a thin layer between http and the Arduino update code. You should do something similar for FTP then we can add generically if needed later. The OCPP module can then also use the new FTP module as needed.

**Comment by matth-x** [comment]
Okay, sounds good. Thanks for the input!
